test/common/filesystem/directory_test.cc

```diff
@@ -0,0 +1,228 @@
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+      {"sub_dir", FileType::Directory},
+      {"file", FileType::Regular},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, false));
+}
+
+TEST(Directory, DirectoryWithSymlinkToFile) {
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test"));
+  const std::string file_path(dir_path + "/" + "file");
+  const std::string link_path(dir_path + "/" + "link");
+  ScopedPathRemover s({link_path, file_path, dir_path});
+  TestUtility::createDirectory(dir_path);
+  { std::ofstream file(file_path); }
+  TestUtility::createSymlink(file_path, link_path);
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+      {"file", FileType::Regular},
+      {"link", FileType::Regular},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, false));
+}
+
+TEST(Directory, DirectoryWithSymlinkToDirectory) {
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test"));
+  const std::string link_dir_path(dir_path + "/" + "link_dir");
+  const std::string child_dir_path(dir_path + "/" + "sub_dir");
+  ScopedPathRemover s({link_dir_path, child_dir_path, dir_path});
+  TestUtility::createDirectory(dir_path);
+  TestUtility::createDirectory(child_dir_path);
+  TestUtility::createSymlink(child_dir_path, link_dir_path);
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+      {"sub_dir", FileType::Directory},
+      {"link_dir", FileType::Directory},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, false));
+}
+
+TEST(Directory, DirectoryWithEmptyDirectory) {
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test"));
+  ScopedPathRemover s({dir_path});
+  TestUtility::createDirectory(dir_path);
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, false));
+}
+
+TEST(DirectoryIteratorImpl, NonExistingDir) {
+  const std::string dir_path("some/non/existing/dir");
+
+#if !defined(WIN32)
+  EXPECT_THROW_WITH_MESSAGE(
+      DirectoryIteratorImpl dir_iterator(dir_path), EnvoyException,
+      fmt::format("unable to open directory {}: No such file or directory", dir_path));
+#else
+  EXPECT_THROW_WITH_MESSAGE(
+      DirectoryIteratorImpl dir_iterator(dir_path), EnvoyException,
+      fmt::format("unable to open directory {}: {}", dir_path, ERROR_PATH_NOT_FOUND));
+#endif
+}
+
+TEST(Directory, DirectoryHasTrailingPathSeparator) {
+#if !defined(WIN32)
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test") + "/");
+#else
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test") + "\\");
+#endif
+  ScopedPathRemover s({dir_path});
+  TestUtility::createDirectory(dir_path);
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
```

Comment:
BTW, do you think we could put some comments or other checks to ensure we never use this code on the data path or with untrusted inputs? I'm kind of terrified of directory traversal attacks and relative paths.

Reply:
Sure, right now it is only used by `DiskLayer::walkDirectory` which I assume only gets trusted inputs (if it didn't, this attack vector would exist even before this change)

By the way, what is the exact meaning of 'data path' in the Envoy context? I've seen it used in conversations and have an idea as to what it means, but would like to be sure I'm on the same page